nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/registry/flow/mapping/NiFiRegistryFlowMapperTest.java

```diff
@@ -0,0 +1,775 @@
+    }
+
+    private void verifyParameterContext(final VersionedParameterContext versionedParameterContext, final ParameterContext parameterContext) {
+        assertEquals(versionedParameterContext.getName(), parameterContext.getName());
+
+        final Collection<Parameter> parameters = parameterContext.getParameters().values();
+        final Set<VersionedParameter> versionedParameters = versionedParameterContext.getParameters();
+        // parameter order is not deterministic - use unique names to map up matching parameters
+        final Iterator<Parameter> parametersIterator = parameters.iterator();
+        while (parametersIterator.hasNext()) {
+            final Parameter parameter = parametersIterator.next();
+            final Iterator<VersionedParameter> versionedParameterIterator = versionedParameters.iterator();
+            while (versionedParameterIterator.hasNext()) {
+                final VersionedParameter versionedParameter = versionedParameterIterator.next();
+                if (versionedParameter.getName().equals(parameter.getDescriptor().getName())) {
+                    verifyParameter(versionedParameter, parameter);
+                    versionedParameterIterator.remove();
+                    break;
+                }
+            }
+        }
+        assertTrue("Failed to match parameters by unique name", versionedParameters.isEmpty());
+
+    }
+
+    private void verifyParameter(final VersionedParameter versionedParameter, final Parameter parameter) {
+        final ParameterDescriptor parameterDescriptor = parameter.getDescriptor();
+
+        assertEquals(versionedParameter.getName(), parameterDescriptor.getName());
+        assertEquals(versionedParameter.getDescription(), parameterDescriptor.getDescription());
+        assertEquals(versionedParameter.isSensitive(), parameterDescriptor.isSensitive());
+        if (parameterDescriptor.isSensitive()) {
+            // verify parameter value is null for sensitive parameters
+            assertNull(versionedParameter.getValue());
+        } else {
+            assertEquals(versionedParameter.getValue(), parameter.getValue());
+        }
+    }
+
+    private ProcessGroup prepareProcessGroup(final int numProcessors, final boolean includeFunnel,final boolean includePorts,
+                                             final boolean includeLabels, final boolean includeVariableRegistry,
+                                             final boolean includeControllerService,
+                                             final ControllerServiceNode externalControllerServiceNode,
+                                             final boolean includeRemoteProcessGroup, final boolean includeVersionControlInfo,
+                                             final List<ProcessGroup> childProcessGroups) {
+        final String processGroupId = UUID.randomUUID().toString();
+        final ProcessGroup processGroup = mock(ProcessGroup.class);
```

Comment:
It looks like there's quite a lot of mocking going on here for the ProcessGroup. There already exists a `MockProcessGroup` class that may simplify this quite a bit.

Reply:
MockProcessGroup was limited in what it was implementing already so I couldn't use it without adding a ton more implementation. Since this mapper exercises almost all the elements of a process group, it was going to take quite a bit to make the MockProcessGroup usable.